Add missing links now that the SRIOV VA has merged (#68)
openstack-k8s-operators/architecture: docs/wa/missing_nncp.md

# Missing NNCPs

> The following command defines the OpenStack control plane and its
> dependent CRs.

Reviewer: ```suggestion
The kustomize command builds and results in the OpenStack control plane definitions and its
dependent Custom Resources (CR).
```

```bash
kustomize build architecture/examples/va/hci > control-plane.yaml
```

> The `control-plane.yaml` file contains CRs for both `NMState` and
> `NodeNetworkConfigurationPolicy` (NNCP). When `oc apply -f` is passed
> this file, OpenShift might try to create the NNCPs while `NMState`
> CRDs are still installing and produce the following message.

Reviewer: ```suggestion
The `control-plane.yaml` file contains CRs for both `NMState` and
`NodeNetworkConfigurationPolicy` (NNCP). When  `oc apply -f control-plane.yaml` is read, OpenShift will try to create the NNCPs while `NMState`
Custom Resource Definitions (CRD) are still installing and produce a message noting that the resource mappings are not found:
```

```
nmstate.nmstate.io/nmstate created
[resource mapping not found for name:
"ostest-master-0" namespace: "openstack" from "control-plane.yaml":
no matches for kind "NodeNetworkConfigurationPolicy" in version "nmstate.io/v1"
ensure CRDs are installed first,
resource mapping not found for name: "ostest-master-1" namespace: "openstack"
from "control-plane.yaml": no matches for kind "NodeNetworkConfigurationPolicy"
in version "nmstate.io/v1"
```

> Retrying `oc apply -f contol-plane.yaml` a few seconds later should
> resolve the problem however.

Reviewer: ```suggestion
Retrying `oc apply -f contol-plane.yaml` a few seconds later is likely to
resolve the problem.
```

## Alternative Approach

> It's also possible to create CR files with less components and wait
> before applying each CR file. E.g. the file `nncp.yaml` would contain
> only `NodeNetworkConfigurationPolicy` CRs and `NMState` and other
> deployment related CRs could exist in another file like
> `deploy.yaml`. The following process may be used to do generate these
> files using kustomize.

Reviewer: ```suggestion
You can create Custom Resource (CR) files with less components and wait
before applying each CR file. For example, the file `nncp.yaml` could contain
only `NodeNetworkConfigurationPolicy` CRs and `NMState` definitions. Other
deployment related CRs could exist in another file such as
`deploy.yaml`. You can generate these files using kustomize.
```

- Edit

> [va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/hci/kustomization.yaml)

Author: URL paths are used here instead of relative paths because it's within the contents of the `docs` directory which can also be used to generate its own [mkdocs](https://www.mkdocs.org) page.

https://github.com/openstack-k8s-operators/architecture/blob/main/docs/contributing/documentation.md

> - Edit
> [va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/hci/kustomization.yaml)
> so that the last two lines contain only the following:
> ```yaml
> components:
> - ../../lib/deploy
> ```
> - Generate a file with only `MetalLB` and `NMState` CRs:
> ```bash
> kustomize build architecture/examples/va/hci > deploy.yaml
> ```
> - Edit
> [va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/hci/kustomization.yaml)
> so that the last two lines contain only the following:
> ```yaml
> components:
> - ../../lib/nncp
> ```
> - Generate a file with only `NNCP` CRs:
> ```bash
> kustomize build architecture/examples/va/hci > nncp.yaml
> ```
> The above process may be continued for each component.
> 
> Note that [va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/hci/kustomization.yaml)
> is not the same file as
> [examples/va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/examples/va/hci/kustomization.yaml).
> `/example/va/hci` is a specific example of a given VA where as
> `/va/hci` is a generic HCI VA that may be customised and shared in
> multiple examples or composed to make a larger VA.
> 
> This process will work for VAs (and DTs) besides HCI, but the paths
> may be different. E.g.
> [examples/va/nfv/sriov/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/examples/va/nfv/sriov/kustomization.yaml)
> differs from
> [va/nfv/sriov/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/nfv/sriov/kustomization.yaml)
> and the later is in an `nfv` subdirectory
> so each component is referred to using
> `- ../../../lib/` instead of `- ../../lib/`.

Reviewer: ```suggestion
- Modify the [va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/hci/kustomization.yaml) file for your environment
so that the last two lines contain the deploy component configuration:
```yaml
components:
- ../../lib/deploy
```
- Generate a file with `MetalLB` and `NMState` CRs:
```bash
kustomize build architecture/examples/va/hci > deploy.yaml
```
- Modify the [va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/hci/kustomization.yaml) file
so that the last two lines contain the NNCP component configuration:
```yaml
components:
- ../../lib/nncp
```
- Generate a file with `NNCP` CRs:
```bash
kustomize build architecture/examples/va/hci > nncp.yaml
```
You can repeat a similar process for each component.

Note that [va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/hci/kustomization.yaml)
is not the same file as
[examples/va/hci/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/examples/va/hci/kustomization.yaml).

The `/example/va/hci` is a specific example of a given verified architecture (VA) where
`/va/hci` is a generic hyperconverged infrastructure (HCI) VA that may be customised and shared in
multiple examples or composed to make a larger VA.

You can repeat the process for VAs (and deployment topologies (DT)) besides HCI, but the paths
may be different. For example,
[examples/va/nfv/sriov/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/examples/va/nfv/sriov/kustomization.yaml)
differs from
[va/nfv/sriov/kustomization.yaml](https://github.com/openstack-k8s-operators/architecture/blob/main/va/nfv/sriov/kustomization.yaml)
and the latter is a subdirectory named `nfv` 
so the relative path of each component is referred to using
`- ../../../lib/` instead of `- ../../lib/`.
```
